core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,228 @@
+// Copyright (C) 2018 - 2019 John A. De Goes. All rights reserved.
+package scalaz.zio.internal
+
+import java.util.concurrent._
+import scala.concurrent.ExecutionContext
+
+/**
+ * An executor is responsible for executing actions. Each action is guaranteed
+ * to begin execution on a fresh stack frame.
+ */
+trait Executor {
+
+  /**
+   * The role the executor is optimized for.
+   */
+  def role: Executor.Role
+
+  /**
+   * The concurrency level of the executor.
+   */
+  def concurrency: Int
+
+  /**
+   * The capacity of the executor.
+   */
+  def capacity: Int
+
+  /**
+   * The number of tasks remaining to be executed.
+   */
+  def size: Int
+
+  /**
+   * The number of tasks that have been enqueued, over all time.
+   */
+  def enqueuedCount: Long
+
+  /**
+   * The number of tasks that have been dequeued, over all time.
+   */
+  def dequeuedCount: Long
+
+  /**
+   * The number of operations a fiber should run before yielding.
+   */
+  def yieldOpCount: Int
+
+  /**
+   * Submits a task for execution.
+   */
+  def submit(runnable: Runnable): Boolean
+
+  /**
+   * Initiates shutdown of the executor.
+   */
+  def shutdown(): Unit
+
+  /**
+   * Views this `Executor` as a Scala `ExecutionContext`.
+   */
+  lazy val asExecutionContext: ExecutionContext =
+    new ExecutionContext {
+      override def execute(r: Runnable): Unit =
+        if (!submit(r)) throw new RejectedExecutionException("Rejected: " + r.toString)
+
+      override def reportFailure(cause: Throwable): Unit =
+        cause.printStackTrace
+    }
+}
+
+object Executor extends Serializable {
+  sealed abstract class Role extends Product with Serializable {
+    def defaultYieldOpCount: Int = this match {
```

Comment:
Is it to avoid memory allocation that this is not a val?
```scala
final case object Unyielding extends Role(Int.MaxValue)
```

Reply:
Fixed in the new version.